<a href="https://colab.research.google.com/github/PradeepKumarM30/MachineLearningtechniques/blob/main/20MAI0030EX2/EX2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [42]:
#Load all the neccesary imports required for the project 
import pandas as pd
import numpy as np


# **DATASET 1 : CRX.DATA**

In [43]:
#Place the data in the dataframe for easy visulaisation of the data and verify other stats of the data
#that are yet to come
dataFrameForcrxData = pd.read_csv('crx.data',header=None)

# **QNO 1 : In the dataset crx.data, name the columns as col1, col2..col16**

In [44]:
#Renaming the columns to required format
colnames = ['col' + str(s) for s in range(1,17)]

In [45]:
#Verify the operation
print(colnames)

['col1', 'col2', 'col3', 'col4', 'col5', 'col6', 'col7', 'col8', 'col9', 'col10', 'col11', 'col12', 'col13', 'col14', 'col15', 'col16']


In [46]:
#reflect the change in the dataframe
dataFrameForcrxData.columns = colnames

In [47]:
#Glimpse of dataframe
dataFrameForcrxData.head()

col1   col2   col3 col4 col5 col6  ... col11  col12 col13  col14  col15 col16
0    b  30.83  0.000    u    g    w  ...     1      f     g  00202      0     +
1    a  58.67  4.460    u    g    q  ...     6      f     g  00043    560     +
2    a  24.50  0.500    u    g    q  ...     0      f     g  00280    824     +
3    b  27.83  1.540    u    g    w  ...     5      t     g  00100      3     +
4    b  20.17  5.625    u    g    w  ...     0      f     s  00120      0     +

[5 rows x 16 columns]

# **QNO2 : Print the last 10 rows of data**

In [48]:
dataFrameForcrxData.tail(10)

col1   col2    col3 col4 col5 col6  ... col11  col12 col13  col14  col15 col16
680    b  19.50   0.290    u    g    k  ...     0      f     g  00280    364     -
681    b  27.83   1.000    y    p    d  ...     0      f     g  00176    537     -
682    b  17.08   3.290    u    g    i  ...     0      t     g  00140      2     -
683    b  36.42   0.750    y    p    d  ...     0      f     g  00240      3     -
684    b  40.58   3.290    u    g    m  ...     0      t     s  00400      0     -
685    b  21.08  10.085    y    p    e  ...     0      f     g  00260      0     -
686    a  22.67   0.750    u    g    c  ...     2      t     g  00200    394     -
687    a  25.25  13.500    y    p   ff  ...     1      t     g  00200      1     -
688    b  17.92   0.205    u    g   aa  ...     0      f     g  00280    750     -
689    b  35.00   3.375    u    g    c  ...     0      t     g  00000      0     -

[10 rows x 16 columns]

# **QNO3 . Replace the ‘?’ with Not-a-Number**

In [49]:
dataFrameForcrxData = dataFrameForcrxData.replace('?',np.nan)

In [50]:
#Exploring the dataset for undersatding the behaviour of the features
ataFrameForcrxData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 690 entries, 0 to 689
Data columns (total 16 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   col1    678 non-null    object 
 1   col2    678 non-null    object 
 2   col3    690 non-null    float64
 3   col4    684 non-null    object 
 4   col5    684 non-null    object 
 5   col6    681 non-null    object 
 6   col7    681 non-null    object 
 7   col8    690 non-null    float64
 8   col9    690 non-null    object 
 9   col10   690 non-null    object 
 10  col11   690 non-null    int64  
 11  col12   690 non-null    object 
 12  col13   690 non-null    object 
 13  col14   677 non-null    object 
 14  col15   690 non-null    int64  
 15  col16   690 non-null    object 
dtypes: float64(2), int64(2), object(12)
memory usage: 86.4+ KB


In [51]:
dataFrameForcrxData['col1'].value_counts()

b    468
a    210
Name: col1, dtype: int64

In [52]:
dataFrameForcrxData['col4'].value_counts()

u    519
y    163
l      2
Name: col4, dtype: int64

In [53]:
dataFrameForcrxData['col5'].value_counts()

g     519
p     163
gg      2
Name: col5, dtype: int64

In [54]:
dataFrameForcrxData['col16'].value_counts()

-    383
+    307
Name: col16, dtype: int64

In [55]:
dataFrameForcrxData.describe()

col3        col8      col11          col15
count  690.000000  690.000000  690.00000     690.000000
mean     4.758725    2.223406    2.40000    1017.385507
std      4.978163    3.346513    4.86294    5210.102598
min      0.000000    0.000000    0.00000       0.000000
25%      1.000000    0.165000    0.00000       0.000000
50%      2.750000    1.000000    0.00000       5.000000
75%      7.207500    2.625000    3.00000     395.500000
max     28.000000   28.500000   67.00000  100000.000000

# **QNO.4 Comment on the datatype of variables**

Comments on the datatypes of the crx data columns



*   All the columns are not null columns 
*   Columns col3 and col8 contains float values also referred to as continous values
*   Col1 contains only categorical values a and b(binary values count 468 and 210 respectively)
*   Similarly COL4 contains only u,y and l values (categorical - three values), col5 contains g,p and gg, col6 with few values (categorical eg.c,q,w,etc)
*   col7,col9,col10,col12,col13 are few other categorical columns which consists of values that contains only values that belong to a particular set with rescept to that column
* The column names are used after the renaming of the column names of the original data set to the custom named column names


# **Q.NO.5 The col16 has + and -, replace them ‘P’ and ‘N’ respectively**

In [56]:
#Enocde the target variables from + and  - to P and N respectively
dataFrameForcrxData['col16'] = dataFrameForcrxData['col16'].map({'+':'P','-':'N'})


In [57]:
#Is all going right? we shall check it out
dataFrameForcrxData

col1   col2    col3 col4 col5 col6  ... col11  col12 col13  col14  col15 col16
0      b  30.83   0.000    u    g    w  ...     1      f     g  00202      0     P
1      a  58.67   4.460    u    g    q  ...     6      f     g  00043    560     P
2      a  24.50   0.500    u    g    q  ...     0      f     g  00280    824     P
3      b  27.83   1.540    u    g    w  ...     5      t     g  00100      3     P
4      b  20.17   5.625    u    g    w  ...     0      f     s  00120      0     P
..   ...    ...     ...  ...  ...  ...  ...   ...    ...   ...    ...    ...   ...
685    b  21.08  10.085    y    p    e  ...     0      f     g  00260      0     N
686    a  22.67   0.750    u    g    c  ...     2      t     g  00200    394     N
687    a  25.25  13.500    y    p   ff  ...     1      t     g  00200      1     N
688    b  17.92   0.205    u    g   aa  ...     0      f     g  00280    750     N
689    b  35.00   3.375    u    g    c  ...     0      t     g  00000      0     N

[690 rows x 16 columns]

# **QNO6. Find and display the number of variables of type ‘Object**

In [58]:
dataFrameForcrxData.dtypes.value_counts()

object     12
float64     2
int64       2
dtype: int64

 **It can be seen that there are 12 variables of the type object. Objects usually contains strings and other datatypes also**

# **DATASET2 : LOAN.CSV**

In [59]:
#Load the dataset
dataFrameForLoanCsv = pd.read_csv('loan.csv')

In [60]:
#A glimpse of data
dataFrameForLoanCsv.head()

customer_id  disbursed_amount  ...  date_last_payment number_credit_lines_12
0            0           23201.5  ...         2016-01-14                    NaN
1            1            7425.0  ...         2016-01-25                    NaN
2            2           11150.0  ...         2014-09-26                    NaN
3            3            7600.0  ...         2016-01-26                    NaN
4            4           31960.0  ...         2016-01-11                    NaN

[5 rows x 14 columns]

In [61]:
#Understanding the dtypes of the each column
dataFrameForLoanCsv.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 14 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   customer_id             10000 non-null  int64  
 1   disbursed_amount        10000 non-null  float64
 2   interest                10000 non-null  float64
 3   market                  10000 non-null  object 
 4   employment              9389 non-null   object 
 5   time_employed           9471 non-null   object 
 6   householder             10000 non-null  object 
 7   income                  10000 non-null  float64
 8   date_issued             10000 non-null  object 
 9   target                  10000 non-null  int64  
 10  loan_purpose            10000 non-null  object 
 11  number_open_accounts    10000 non-null  float64
 12  date_last_payment       10000 non-null  object 
 13  number_credit_lines_12  238 non-null    float64
dtypes: float64(5), int64(2), object(7)
memo

disbursed_amount,income,intrest,number_open_accounts and number_credit_lines are found to have continous values. We shall choose two among them for the QNO1

# **QNO1:Display the mean of any two variables with continuous values**

In [62]:
dataFrameForLoanCsv['disbursed_amount'].mean()

14132.2755

In [63]:
dataFrameForLoanCsv['number_open_accounts'].mean()

12.5163

In [65]:
dataFrameForLoanCsv.dtypes.value_counts()

object     7
float64    5
int64      2
dtype: int64

# **QNO2:Print the number of discrete variables**

 **As far as i understand, discrete variables are those which contain values from a particular set of values. So i have proceeded with the understanding that the discrete and categorical are the same. If numbers like 2,3,4,.. are discrete, they are actually not as they are nothing but 2.00000..,3.00000..,4.0000..,etc. making it a continous data. With this understanding i have calculated the upcoming questions.**

In [132]:
#Remove the number of numeric data columns from the total number of columns
cols = dataFrameForLoanCsv.columns

num_cols = dataFrameForLoanCsv._get_numeric_data().columns

print('The number of discrete variables ',len(list(set(cols) - set(num_cols))),'and they are ',list(set(cols) - set(num_cols)))


The number of discrete variables  7 and they are  ['loan_purpose', 'date_issued', 'householder', 'date_last_payment', 'time_employed', 'market', 'employment']


In [67]:
dataFrameForLoanCsv['market'].value_counts()

B    2846
C    2730
A    1673
D    1577
E    1174
Name: market, dtype: int64

Discrete variables a.k.a categorized variables 

# **QNO3:Display the unique values of two variables with discrete values**

In [68]:
dataFrameForLoanCsv['market'].unique()

array(['C', 'B', 'A', 'E', 'D'], dtype=object)

In [69]:
dataFrameForLoanCsv['householder'].unique()

array(['RENT', 'OWNER', 'MORTGAGE'], dtype=object)

In [73]:
dataFrameForLoanCsv.dtypes

customer_id                 int64
disbursed_amount          float64
interest                  float64
market                     object
employment                 object
time_employed              object
householder                object
income                    float64
date_issued                object
target                      int64
loan_purpose               object
number_open_accounts      float64
date_last_payment          object
number_credit_lines_12    float64
dtype: object

# **QNO4: Display the Month with most of loans issued date**

In [76]:
#Firstly converting the object into date format
dataFrameForLoanCsv['date_issued'] =  pd.to_datetime(dataFrameForLoanCsv['date_issued'])


In [80]:
#Checking if all is well!
dataFrameForLoanCsv.dtypes

customer_id                        int64
disbursed_amount                 float64
interest                         float64
market                            object
employment                        object
time_employed                     object
householder                       object
income                           float64
date_issued               datetime64[ns]
target                             int64
loan_purpose                      object
number_open_accounts             float64
date_last_payment                 object
number_credit_lines_12           float64
dtype: object

In [82]:
#extract the month from the date column
dataFrameForLoanCsv['month']= dataFrameForLoanCsv['date_issued'].dt.month 


In [83]:
#dataframe with month column added, can be remove once the processing is done
dataFrameForLoanCsv

customer_id  disbursed_amount  ...  number_credit_lines_12 month
0               0           23201.5  ...                     NaN     6
1               1            7425.0  ...                     NaN     5
2               2           11150.0  ...                     NaN    10
3               3            7600.0  ...                     NaN     8
4               4           31960.0  ...                     NaN     7
...           ...               ...  ...                     ...   ...
9995         9995           23750.0  ...                     NaN     1
9996         9996           11880.0  ...                     NaN     3
9997         9997           19950.0  ...                     NaN     4
9998         9998            4850.0  ...                     NaN    11
9999         9999           25584.0  ...                     NaN     1

[10000 rows x 15 columns]

In [103]:
#Finding the month with maximum frequency and encoding it with the help of dictionary
months_ref = {'1':'Jan','2':'Feb','3':'Mar','4':'Apr','5':'May','6':'Jun','7':'Jul','8':'Aug','9':'Sept','10':'Oct','11':'Nov','12':'Dec'}
print(months_ref[str(dataFrameForLoanCsv.month.mode()[0])])

Oct


In [105]:

dataFrameForLoanCsv.month.mode()

0    10
dtype: int64

In [117]:
#Some extractions before jumping to the QNO4
Teacher = dataFrameForLoanCsv['employment'] == "Teacher"

Owner = dataFrameForLoanCsv['householder'] == "OWNER"

dataFrameForLoanCsv[Teacher & Owner]

customer_id  disbursed_amount  ...  number_credit_lines_12 month
71             71          10230.00  ...                     NaN     2
85             85           3713.00  ...                     NaN     7
171           171          19646.25  ...                     NaN     9
672           672          15200.00  ...                     NaN    11
1024         1024          12144.00  ...                     NaN     4
...           ...               ...  ...                     ...   ...
9154         9154           9900.00  ...                     NaN    12
9172         9172          22080.00  ...                     NaN     8
9433         9433           4750.00  ...                     NaN     7
9684         9684           6650.00  ...                     NaN     4
9769         9769          16005.00  ...                     NaN     9

[69 rows x 15 columns]

# **QNO5:Display the count of ‘Teacher’ who are ‘Owners’**

In [119]:
dataFrameForLoanCsv[Teacher & Owner].shape[0]

69

In [120]:
dataFrameForLoanCsv

customer_id  disbursed_amount  ...  number_credit_lines_12 month
0               0           23201.5  ...                     NaN     6
1               1            7425.0  ...                     NaN     5
2               2           11150.0  ...                     NaN    10
3               3            7600.0  ...                     NaN     8
4               4           31960.0  ...                     NaN     7
...           ...               ...  ...                     ...   ...
9995         9995           23750.0  ...                     NaN     1
9996         9996           11880.0  ...                     NaN     3
9997         9997           19950.0  ...                     NaN     4
9998         9998            4850.0  ...                     NaN    11
9999         9999           25584.0  ...                     NaN     1

[10000 rows x 15 columns]

# **QNO6:Display the ‘Employment’ of customers who mostly ‘Rent**

In [129]:
#Firstly extract the rent rows and then use that dataframe to find the employments of the people under the rent 
#householder category
Rent = dataFrameForLoanCsv['householder'] == "RENT"
newDataFrame = dataFrameForLoanCsv[Rent]
newDataFrame['employment'].value_counts()

Civil Servant         371
Teacher               371
Bus driver            360
Nurse                 358
Secretary             355
Dentist               355
Other                 353
Statistician          342
Accountant            322
Taxi driver           316
Software developer    315
Name: employment, dtype: int64

In [130]:
newDataFrame

customer_id  disbursed_amount  ...  number_credit_lines_12 month
0               0           23201.5  ...                     NaN     6
2               2           11150.0  ...                     NaN    10
3               3            7600.0  ...                     NaN     8
4               4           31960.0  ...                     NaN     7
6               6           27900.0  ...                     NaN     9
...           ...               ...  ...                     ...   ...
9991         9991           20580.0  ...                     NaN     8
9992         9992            1470.0  ...                     1.0    12
9994         9994            5520.0  ...                     NaN    11
9996         9996           11880.0  ...                     NaN     3
9998         9998            4850.0  ...                     NaN    11

[4055 rows x 15 columns]